apps/web/utils/discord-markdown.ts

```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
 import { db } from '@nextjs-forum/db/node'
 import { getCanonicalPostUrl } from './urls'
 import { sanitizeText } from 'simple-markdown'
+import LRUCache from 'lru-cache'
+import { unstable_cache } from 'next/cache'
 
-export const parseDiscordMessage = async (content: string) => {
-  // Mentions handling
-  const memberIds = new Set<string>()
-  const channelIds = new Set<string>()
-
-  // convert https://discord.com/channels/<guild_id>/<post_id> to https://nextjs-forum.com/posts/<post_id>
-  // or https://discord.com/channels/<guild_id>/<channel_id>/<message_id> to https://nextjs-forum.com/posts/<post_id>#message-<message_id>
-  const regex = /https:\/\/discord\.com\/channels\/(?<guild>\d+)\/(?<channel>\d+)(\/(?<message>\d+))?/g;
-  const postIds = new Set<string>(Array.from(content.matchAll(regex), (m) => m.groups?.channel ?? ''))
-
-  const posts =
-    postIds.size > 0
-      ? await db
-        .selectFrom('posts')
-        .select(['snowflakeId', 'title'])
-        .where('snowflakeId', 'in', Array.from(postIds))
-        .execute()
-      : []
-
-  // Parse the content again, this time replacing the link
-  content = content.replace(regex, (match, guild, channel, _, message) => {
-    const post = posts.find((p) => p.snowflakeId === channel)
+
+interface UserCache {
+  snowflakeId: string
+  username: string
+}
+
+interface ChannelCache {
+  snowflakeId: string
+  name: string
+}
+
+interface PostCache {
+  snowflakeId: string
+  title: string
+}
+
+const QUERY_REVALIDATE_TIME = 60 // 1 minute
+
+const fetchUser = (userId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('users')
+      .select(['snowflakeId', 'username'])
+      .where('snowflakeId', '=', userId)
+      .executeTakeFirst()
+  ) || null
```

Comment:
`executeTakeFirst` can return `undefined` if the row wasn't found so we could use it instead of changing it to `null`. This suggestion applies to the other functions

Reply:
The reason for this was that the types get messed up later on:
* `internalLink` wants `{ snowflakeId: string; title: string; }[]` but is given `({ snowflakeId: string; title: string; } | undefined)[]`
* less of an issue: the mentions finding just needs an extra question mark `.find((u) => u?.`

But your right, it is a bit unnecessary...

Reply:
Wait, now I remember the main issue was when the mention can't be found (dev and normal mode have this):
```js
@nextjs-forum/web:dev: - error ..\node_modules\next\dist\server\web\spec-extension\unstable-cache.js (84:35) @ parse
@nextjs-forum/web:dev: - error SyntaxError: "undefined" is not valid JSON
@nextjs-forum/web:dev:     at JSON.parse (<anonymous>)
@nextjs-forum/web:dev:     at async Promise.all (index 1)
@nextjs-forum/web:dev:     at async Promise.all (index 2)
@nextjs-forum/web:dev:     at async eval (./utils/discord-markdown.ts:66:38)
@nextjs-forum/web:dev:     at async eval (./utils/discord-markdown.ts:79:40)
@nextjs-forum/web:dev:     at async Post (./app/post/[id]/page.tsx:163:23)
@nextjs-forum/web:dev: digest: "3742489035"
@nextjs-forum/web:dev: null
```

Reply:
Ahh I see, `unstable_cache` is serializing the data to JSON so we can't use `undefined`. In this case using `null` is fine!